index.bs

```diff
@@ -1046,101 +1010,467 @@ BigIntValue = {
+  RegExpLocalValue //
+  SetLocalValue //
 }
 
-ArrayValue = {
+ListLocalValue = [*LocalValue];
+
+ArrayLocalValue = {
   type: "array",
-  objectId: ObjectId,
-  value?: ListValue,
+  value: ListLocalValue,
+}
+
+DateLocalValue = {
+  type: "date",
+  value: text
 }
 
-ObjectValue = {
+MappingLocalValue = [*[(LocalValue / text), LocalValue]];
+
+MapLocalValue = {
+  type: "map",
+  value: MappingLocalValue,
+}
+
+ObjectLocalValue = {
   type: "object",
-  objectId: ObjectId,
-  value?: MappingValue,
+  value: MappingLocalValue,
 }
 
-FunctionValue = {
-  type: "function",
+RegExpLocalValue = {
+  type: "regexp",
+  pattern: text,
+  ?flags: text
+}
+
+SetLocalValue = {
+  type: "set",
+  value: ListLocalValue
+}
+```
+
+<div algorithm>
+
+To <dfn>deserialize key-value list</dfn> given a |serialized key-value list|:
+
+1. Let |deserialized key-value list| be a new list.
+
+1. For each |serialized key-value| in the |serialized key-value list|:
+
+   1. If [=list/size=] of |serialized key-value| is not 2,
+      return [=error=] with [=error code=] [=invalid argument=].
+
+   1. Let |serialized key| be |serialized key-value|[0].
+
+   1. If |serialized key| is a <code>string</code>,
+      let |deserialized key| be |serialized key|.
+
+   1. Otherwise let |deserialized key| be result of [=trying=] to [=deserialize local value=]
+      with |serialized key|.
+
+   1. Let |serialized value| be |serialized key-value|[1].
+
+   1. Let |deserialized value| be result of [=trying=] to [=deserialize local value=] with
+      |serialized value|.
+
+   1. Append a <code>new [=Array=](|deserialized key|, |deserialized value|)</code>
+      to |deserialized key-value list|.
+
+1. Return [=success=] with data |deserialized key-value list|.
+
+</div>
+
+<div algorithm>
+
+Issue: TODO distinguish between `List` and `Array`.
+
+To <dfn>deserialize value list</dfn> given a |serialized value list|:
+
+1. Let |deserialized values| be a new list.
+
+1. For each |serialized value| in the |serialized value list|:
+
+   1. Let |deserialized value| be result of [=trying=] to [=deserialize local value=] with
+      |serialized value|.
+
+   1. Append |deserialized value| to |deserialized values|;
+
+1. Return [=success=] with data |deserialized values|.
+
+</div>
+
+<div algorithm>
+
+To <dfn>deserialize local value</dfn> given a |local protocol value|:
+
+1. If |local protocol value| matches the <code>PrimitiveProtocolValue</code> production, return
+   [=deserialize primitive protocol value=] with |local protocol value|.
+
+1. Let |type| be the value of the <code>type</code> field of |local protocol value| or undefined if
+   no such a field.
+
+1. Let |value| be the value of the <code>value</code> field of |local protocol value| or undefined
+   if no such a field.
+
+1. In the following list of conditions and associated steps, run the first set of steps for which
+   the associated condition is true. If this throws exception, return [=error=] with [=error code=]
```

Comment:
Again, I think we need to handle failure in each case.

Reply:
done. Exception is expected only in Regex parsing.